### PR TITLE
Allow Weak::as_ptr and friends for unsized T

### PR DESCRIPTION
**Pull request by CAD97**
Relaxes `impl<T> Weak<T>` to `impl<T: ?Sized> Weak<T>` for the methods `rc::Weak::as_ptr`, `into_raw`, and `from_raw`.

Follow-up to #73845, which did most of the impl work to make these functions work for `T: ?Sized`.

We still have to adjust the implementation of `Weak::from_raw` here, however, because I missed a use of `ptr.is_null()` previously. This check was necessary when `into`/`from_raw` were first implemented, as `into_raw` returned `ptr::null()` for dangling weak. However, we now just (wrapping) offset dangling weaks' pointers the same as nondangling weak, so the null check is no longer necessary (or even hit). (I can submit just 17a928f as a separate PR if desired.)

As a nice side effect, moves the `fn is_dangling` definition closer to `Weak::new`, which creates the dangling weak.

This technically stabilizes that "something like `align_of_val_raw`" is possible to do. However, I believe the part of the functionality required by these methods here -- specifically, getting the alignment of a pointee from a pointer where it may be dangling iff the pointee is `Sized` -- is uncontroversial enough to stabilize these methods without a way to implement them on stable Rust.

r? @RalfJung, who reviewed #73845.

ATTN: This changes (relaxes) the (input) generic bounds on stable fn!

### CODE REVIEW

File: src/liballoc/rc.rs
Comment (RalfJung) [inline src/liballoc/rc.rs:1799]: Ah so this is from back when we used NULL as sentinel value for dangling `Weak`?

**Comment by RalfJung** [comment]
I assume the plan is to do the same for `Arc` once we are done with review?

File: src/liballoc/rc.rs
Comment (CAD97) [reply src/liballoc/rc.rs:1799]: Yep, exactly.

**Comment by CAD97** [comment]
Yes.

... Though to be completely honest, I had convinced myself (somehow) that the sync versions already supported unsized values and didn't need to be tweaked. Obviously that's not true, so I don't know _why_ I thought that...

**Comment by RalfJung** [comment]
The code changes look good to me, but I'd like to see unit tests added that use all of the newly `T: ?Sized` with an actually unsized type.

**Comment by CAD97** [comment]
Translated the existing into/from raw tests to also do the same for `Weak`. I still cannot get the Rust test harness to cmake properly on my machine, so I'm relying on CI to get test results here.

**Review by RalfJung — APPROVED** [review]
:+1: from a code perspective.

@rust-lang/libs I think this needs FCP for being an insta-stable change.

**Comment by Amanieu** [comment]
@rfcbot fcp merge

`Weak::from_raw`, `Weak::into_raw`, `Weak::as_raw` now support pointers to DSTs.

**Comment by dtolnay** [comment]
I think I am going to restart this with T-lang on here as well. :/

@rfcbot fcp cancel

**Comment by dtolnay** [comment]
@rust-lang/libs: This makes the following usable with `T: ?Sized`:

- https://doc.rust-lang.org/nightly/alloc/rc/struct.Weak.html#method.as_ptr
- https://doc.rust-lang.org/nightly/alloc/rc/struct.Weak.html#method.into_raw
- https://doc.rust-lang.org/nightly/alloc/rc/struct.Weak.html#method.from_raw
- https://doc.rust-lang.org/nightly/alloc/sync/struct.Weak.html#method.as_ptr
- https://doc.rust-lang.org/nightly/alloc/sync/struct.Weak.html#method.into_raw
- https://doc.rust-lang.org/nightly/alloc/sync/struct.Weak.html#method.from_raw

@rust-lang/lang: Calling attention to the following callout in the PR description:

> This technically stabilizes that "something like `size_of_val_raw`" is possible to do. However, I believe the part of the functionality required by these methods here -- specifically, getting the size of a pointee from a pointer where it may be dangling iff the pointee is `Sized` -- is uncontroversial enough to stabilize these methods without a way to implement them on stable Rust.

(From skimming the implementation I think they might have meant `align_of_val_raw` not `size_of_val_raw`, or maybe both? @CAD97 could you confirm how `size_of_val_raw` is involved in the implementation?)

Here is the code in question:

https://github.com/rust-lang/rust/blob/f244725afda8e124705356f40f75f15ea3af81d3/src/liballoc/rc.rs#L2121-L2137

Effectively this stabilization is committing to the following semantic being implementable internally to std:

```rust
unsafe fn align_of_val_raw<T: ?Sized>(val: *const T) -> usize {
    if T is Sized {
        mem::align_of::<T>()
    } else {
        mem::align_of_val::<T>(&*val)
    }
}
```

Right now we expose the same thing unstably in https://doc.rust-lang.org/nightly/std/mem/fn.align_of_val_raw.html which is tracked by https://github.com/rust-lang/rust/issues/69835.

@rfcbot fcp merge

**Comment by CAD97** [comment]
> meant `align_of_val_raw` not `size_of_val_raw`

Ah, yes, only `align_of_val_raw` is actually needed here. I've chunked the two into the same "feature" in my head, thus the mixup 😅

**Comment by nikomatsakis** [comment]
Notably, [`align_of_val`](https://doc.rust-lang.org/std/mem/fn.align_of_val.html) is already possible and stable, so the only diff here is saying that it can be done from a *raw pointer* and not just a reference, right? This doesn't seem like much of anything at all to me.

**Comment by nikomatsakis** [comment]
Removing nominated label as this was discussed in today's @rust-lang/lang meeting. No concerns were raised.

**Comment by CAD97** [comment]
Rebased the big library move.

I don't know if/when to re-ping team members for pre-FCP-merge checkboxes; we're waiting on those.

**Comment by RalfJung** [comment]
@cramertj @joshtriplett @pnkfelix @withoutboats this FCP is waiting for your checkbox :)

**Comment by RalfJung** [comment]
@cramertj @joshtriplett @pnkfelix @withoutboats FCP reminder :)

**Comment by CAD97** [comment]
#76637 was a rollup; conflict looks like it was from #76530, which is reasonable.

Rebased. FCP is complete; we're just waiting on an r+ now. The sooner this is merged the better; I'd like to replace the use of `set_data_ptr` with [`<*const T>::set_ptr_value`](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.set_ptr_value) now that it's available but that likely conflicts with this again, so I'm waiting until this is merged to do that change.

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

**Comment by RalfJung** [comment]
Sorry, I didn't realize this was waiting for me. I saw the FCP end but had forgotten that I am assigned.^^

File: library/alloc/src/rc.rs
Comment (RalfJung) [inline library/alloc/src/rc.rs:1863]: "iff" isn't right here... the pointer is not "dangling if and only if `T` is sized". That would mean that `T: Sized` implies a dangling pointer.

I think a plain "if" is correct here.

File: library/alloc/src/sync.rs
Comment (RalfJung) [inline library/alloc/src/sync.rs:1646]: Same problem with the iff.

**Review by RalfJung — COMMENTED** [review]
r=me with the comments fixed

**Comment by Amanieu** [comment]
@bors r-

Oops didn't notice @RalfJung requested changes.

File: library/alloc/src/rc.rs
Comment (CAD97) [reply library/alloc/src/rc.rs:1863]: The actual clause is "(may be dangling) iff (T is Sized)", but just regular if is probably fine here either way. I'll change it soon.

-----

In any case, the desired semantics is that (T is Sized) is a necessary but not sufficient precondition for (pointer is dangling). (It is, however, necessary and sufficient for (pointer may dangle).)

I'm too tired at the moment to remember precisely which english is the "correct" way to encode this, probably "only if," if "iff" is not quite right and I felt the reason to use "iff" in the first place.

Just "if" is correct informally though so I'll just change it to that.

Comment (RalfJung) [reply library/alloc/src/rc.rs:1863]: @CAD97 from what I can see, this change is still outstanding.

**Comment by RalfJung** [comment]
@CAD97 I'll just use GH suggestions to change "iff" to "if" and land that.

File: library/alloc/src/rc.rs
Comment (RalfJung) [inline library/alloc/src/rc.rs:1863]:
```suggestion
        // SAFETY: we use wrapping_offset here because the pointer may be dangling (but only if T: Sized).
```

File: library/alloc/src/sync.rs
Comment (RalfJung) [inline library/alloc/src/sync.rs:1646]:
```suggestion
        // SAFETY: we use wrapping_offset here because the pointer may be dangling (but only if T: Sized)
```

**Comment by CAD97** [comment]
Sorry I wasn't able to make the change myself; I've "just" (August... wow time is fast) started grad school and that's eating up all my time now 😅

**Comment by RalfJung** [comment]
No worries, and enjoy grad school. :)

**Pull request merged**
